### ISSUE TITLE
Vault OIDC/JWT bound_claims does not recognize a map

### ISSUE THREAD

**Issue by christopherbloghett-disney** [hashicorp/vault-plugin-auth-jwt#68]
When writing back a role and submitting a json bound claim, it can not writer saying a few different errors. Like recieved a string when expecting a map, or needs to be key value pair. Currently using Docker vault 1.2.1. 

Different Bound_claims I have tried. Based on the very few examples I can find online.

vault write auth/oidc/role/testadmins \
>     user_claim=sub \
>     bound_claims={"roles": ["testadmins", "world"]} \
>     group_claims=testadmins \
>     bound_audiences=vault.com \
>     allowed_redirect_uris=https://vault.com/ui/vault/auth/oidc/oidc/callback \
>     allowed_redirect_uris=https://vault.com/oidc/callback \
>     policies=admins \
>     verbose_oidc_logging=true \
>     ttl=1h
Failed to parse K=V data: invalid key/value pair "[testadmins,": format must be key=value

vault write auth/oidc/role/testadmins \
>     user_claim=sub \
>     bound_claims="roles": ["testadmins", "world"] \
>     group_claims=testadmins \
>     bound_audiences=vault.com \
>     allowed_redirect_uris=https://vault.com/ui/vault/auth/oidc/oidc/callback \
>     allowed_redirect_uris=https://vault.com/oidc/callback \
>     policies=admins \
>     verbose_oidc_logging=true \
>     ttl=1h
Failed to parse K=V data: invalid key/value pair "[testadmins,": format must be key=value

vault write auth/oidc/role/testadmins \
>     user_claim=sub \
>     bound_claims=roles: [testadmins= world] \
>     group_claims=testadmins \
>     bound_audiences=vault.com \
>     allowed_redirect_uris=https://vault.com/ui/vault/auth/oidc/oidc/callback \
>     allowed_redirect_uris=https://vault.com/oidc/callback \
>     policies=admins \
>     verbose_oidc_logging=true \
>     ttl=1h
Failed to parse K=V data: invalid key/value pair "world]": format must be key=value

vault write auth/oidc/role/testadmins \
>     user_claim=sub \
>     bound_claims=roles: [testadmins: world] \
>     group_claims=testadmins \
>     bound_audiences=vault.com \
>     allowed_redirect_uris=https://vault.com/ui/vault/auth/oidc/oidc/callback \
>     allowed_redirect_uris=https://vault.com/oidc/callback \
>     policies=admins \
>     verbose_oidc_logging=true \
>     ttl=1h
Failed to parse K=V data: invalid key/value pair "[testadmins:": format must be key=value

**Expected behavior**
to setup the role with the bound claim.

**Environment:**
* Vault Server Version (retrieve with `vault status`): 1.2.1
* Vault CLI Version (retrieve with `vault version`): v1.2.1
* Server Operating System/Architecture: docker container

**Comment by kalafut** [comment]
Hi @christopherbloghett-disney. What you're seeing is a CLI trait and isn't specific to the JWT plugin. The CLI parameter parsing works well for simple K/V settings and lists, but becomes cumbersome when trying to send in complex data structures like maps. Furthermore, the shell itself can start fighting inline JSON. I think you'll find it much simpler to send the whole configration in as a single JSON blob, e.g.:
```
vault write auth/oidc/role/test -<<EOF
{
  "allowed_redirect_uris": ["abc"],
  "user_claim":"email",
  "groups_claim":"groups",
  "policies":"web",
  "bound_claims": {
    "size": "large"
  }
}
EOF
```

or   `vault write auth/oidc/role/test @my_config.json`

I'll add a note about this to the docs. These more complex configs are common to OIDC.

**Comment by shotah** [comment]
Fine to do this, but that isn't anywhere in the documentation and is definitely a work around.

**Comment by christopherbloghett-disney** [comment]
@kalafut I originally submitted this to Vault because of the CLI. It won't even do K=V pairs via the CLI. 

I have tested the json submission and its working to populate the values. This should close out my immediate need, but this is still a rather large parsing bug out of the CLI.

**Comment by kalafut** [comment]
@christopherbloghett-disney Yeah I realized after the triage this was beyond JWT.

>It won't even do K=V pairs via the CLI.

Do you mean specifically for `bound_claims`, or in other ways too? It will work for `claim_mappings` because the internal type there is a "K/V Pairs" (i.e. string:string) which should be working fine. `bound_claims` is a Map type, so the value can be an arbitrary structure and that's what difficult to put on the CLI today. Maps are used rarely (SSH is the only other place that comes to mind), so this limitation isn't too common.

I'm going to check a few things about Map support on the CLI and will add a Vault issue depending on what I find.

**Comment by shotah** [comment]
Bound claims specifically are the issue. That is why I made the secondary account post about the claim-mappings as they worked as documented.

**Comment by jefferai** [comment]
@christopherbloghett-disney There are two issues here.

One is that you're specifying a KV value that isn't a single value but that the shell is instead interpreting as multiple values due to whitespace. This isn't a Vault CLI issue so much as behavior of your shell. You can fix this with single quoting:

> bound_claims={"roles": ["testadmins", "world"]} \

becomes

> bound_claims='{"roles": ["testadmins", "world"]}' \

At this point you'll get the expected error:

> * error converting input {"roles": ["testadmins", "world"]} for field "bound_claims": '' expected a map, got 'string'

@kalafut We may want to think about enhancing TypeMap to, if it detects a string, attempt to JSON decode the string as a `map[string]interface{}`.

**Comment by christopherbloghett-disney** [comment]
@jefferai yeah, sorry I copied later renditions of my tests, but yes long story short you are unable to pass it in as a variable.

**Comment by llamasoft** [comment]
I just wanted to share my findings regarding this.  What caused me to doubt if passing maps via the CLI was possible was that while the `claim_mappings='{"foo"="bar"}'` argument format _works_, the generated cURL request will have incorrect JSON.

Example:
```
$ vault write -output-curl-string auth/oidc/role/demo claim_mappings='{"email"="name"}'
curl -X PUT -H "X-Vault-Token: $(vault print token)" -d '{"claim_mappings":"{\"email\"=\"name\"}"}' http://localhost:8200/v1/auth/oidc/role/demo
```

You'll note that it has `claim_mappings` as a _string_ containing the text `{"email"="name"}`.  
It just so happens that the Vault server knows what you meant and will parse it correctly for you.

If I could suggest an alternative mapping format for CLI arguments, it would be `var[key]=value`.  Example:
```
vault write auth/oidc/role/demo \
  user_claim=sub \
  claim_mappings["email"]="name" \
  claim_mappings["hd"]="domain" \
  ...
```

It's clear, unambiguous, and unlikely to cause conflicts for any reasonable usage.

**Comment by jefferai** [comment]
The problem is that those are also valid _non-map_ arguments to plenty of potential endpoints in Vault. As a simple example, KV, which at its root stores a JSON object. Should that CLI format map to:
```json
{
    "claim_mappings[\"email\"]": "name"
}
```
or
```json
{
    "claim_mappings": {
        "email": "name"
    }
}
```
We can't know this at the CLI level. So to be nice to CLI users when there is an endpoint that takes in maps or arrays we allow various string-based syntaxes that we then parse. For instance a common pattern is `foo=a,b,c` which for an endpoint that needs a list will automatically be parsed (server side, not CLI side) to `foo = ["a", "b", "c"]`. Only the server/plugin knows when it's appropriate to do that, though. Internally this API endpoint takes in a "TypeKVPairs" where we expect what comes in to not be valid JSON but to actually be `key=value` format.

As for `-output-curl-string` what comes out is indeed a stringified version of the CLI flag but that's accepted in this case for the reasons noted above even though it's not a JSON map...it's because we're not looking for a JSON map, we're looking for KV pairs.

**Comment by stefan-zh** [comment]
I struggled to pass a `TypeMap` CLI argument to a backend endpoint in my project and that's how I found this thread. While passing a `TypeKVPairs` argument works fine like this `vault write auth/<backend>/<path> user="name=john" user="age=23"`, I didn't have any luck getting a `TypeMap` argument to work on the command line, but instead I always got this error: 
```error converting input for field "user": '' expected a map, got 'string'```

@kalafut 's answer above helped me and then I found these examples in the docs https://www.vaultproject.io/docs/commands#stdin:
```sh
echo -n '{"user": {"name": "john", "age": 23}}' | vault write auth/<backend>/<path> -
```
This format helps me because I can test various scenarios inline without having to use a heredoc or reference a file on disk.

**Comment by odu14ick** [comment]
I would just like to add that his renders the CLI that is built in Vault UI completely useless in case you need to use bound_claims (since this CLI only accepts `vault` commands and you can't even echo there.

For instance if you need to create a role for gitlab jwt auth like I do.

Moreover, since there's also no way to create said role from the UI itself - it brings up a question of what is the point of said CLI if not to work around the UI's limitations?

**Comment by fairclothjm** [comment]
This is not specific to this plugin. This is a CLI limitation and the workaround (to write the entire configuration as a single JSON object) is documented here: https://developer.hashicorp.com/vault/docs/auth/jwt#oidc-configuration-troubleshooting

```
vault write auth/oidc/role/demo -<<EOF
{
  "user_claim": "sub",
  "bound_audiences": "abc123",
  "role_type": "oidc",
  "policies": "demo",
  "ttl": "1h",
  "bound_claims": { "groups": ["mygroup/mysubgroup"] }
}
EOF
```
